File: agent_lifecycle_protocol/decommission.py
"""Decommission protocol — apoptosis and necrosis (Section 9).

Implements graceful (apoptosis) and emergency (necrosis) decommission
with estate management, credential revocation tracking, and
counterparty notification.
"""

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from .schema import (
    DECOMMISSION_REASONS,
    CounterpartyNotification,
    Initiator,
    LifecycleEvent,
    _now_iso,
)
from .lifecycle import LifecycleError, LifecycleManager


@dataclass
class CredentialRevocation:
    """Tracks credential revocation during decommission (Section 9.2)."""
    api_keys_revoked: bool = False
    oauth_tokens_invalidated: bool = False
    service_accounts_deleted: bool = False
    trust_relationships_terminated: bool = False
    certificates_revoked: bool = False
    identity_key_archived: bool = False

    @property
    def all_revoked(self) -> bool:
        return (
            self.api_keys_revoked
            and self.oauth_tokens_invalidated
            and self.service_accounts_deleted
            and self.trust_relationships_terminated
            and self.certificates_revoked
            and self.identity_key_archived
        )

    def to_dict(self) -> Dict[str, Any]:
        return {
            "all_api_keys_revoked": self.api_keys_revoked,
            "all_oauth_tokens_invalidated": self.oauth_tokens_invalidated,
            "all_service_accounts_deleted": self.service_accounts_deleted,
            "all_trust_relationships_terminated": self.trust_relationships_terminated,
            "all_certificates_revoked": self.certificates_revoked,
            "identity_key_archived": self.identity_key_archived,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "CredentialRevocation":
        return cls(
            api_keys_revoked=d.get("all_api_keys_revoked", False),
            oauth_tokens_invalidated=d.get("all_oauth_tokens_invalidated", False),
            service_accounts_deleted=d.get("all_service_accounts_deleted", False),
            trust_relationships_terminated=d.get("all_trust_relationships_terminated", False),
            certificates_revoked=d.get("all_certificates_revoked", False),
            identity_key_archived=d.get("identity_key_archived", False),
        )


@dataclass
class DataDisposition:
    """How agent data is handled at decommission (Section 9.2)."""
    operational_logs: str = "archived_90_days"
    memory_state: str = "purged"
    coc_chain: str = "sealed_permanent"
    knowledge_artifacts: str = "transferred_to_fleet"

    def to_dict(self) -> Dict[str, Any]:
        return {
            "operational_logs": self.operational_logs,
            "memory_state": self.memory_state,
            "coc_chain": self.coc_chain,
            "knowledge_artifacts": self.knowledge_artifacts,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "DataDisposition":
        return cls(
            operational_logs=d.get("operational_logs", "archived_90_days"),
            memory_state=d.get("memory_state", "purged"),
            coc_chain=d.get("coc_chain", "sealed_permanent"),
            knowledge_artifacts=d.get("knowledge_artifacts", "transferred_to_fleet"),
        )


@dataclass
class EstateDisposition:
    """Full estate disposition for decommission (Section 5.7)."""
    obligations_transferred: int = 0
    obligations_terminated: int = 0
    credentials: CredentialRevocation = field(
        default_factory=CredentialRevocation
    )
    data: DataDisposition = field(default_factory=DataDisposition)
    counterparties_notified: bool = False
    fleet_coordinator_notified: bool = False

    def to_dict(self) -> Dict[str, Any]:
        return {
            "obligations": {
                "transferred": self.obligations_transferred,
                "terminated": self.obligations_terminated,
            },
            "credentials": self.credentials.to_dict(),
            "data": self.data.to_dict(),
            "notifications": {
                "counterparties_notified": self.counterparties_notified,
                "fleet_coordinator_notified": self.fleet_coordinator_notified,
            },
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "EstateDisposition":
        obligations = d.get("obligations", {})
        notif = d.get("notifications", {})
        return cls(
            obligations_transferred=obligations.get("transferred", 0),
            obligations_terminated=obligations.get("terminated", 0),
            credentials=CredentialRevocation.from_dict(d.get("credentials", {})),
            data=DataDisposition.from_dict(d.get("data", {})),
            counterparties_notified=notif.get("counterparties_notified", False),
            fleet_coordinator_notified=notif.get("fleet_coordinator_notified", False),
        )


@dataclass
class DecommissionPlan:
    """A decommission plan (Section 9.2)."""
    agent_id: str = ""
    reason: str = "end_of_life"
    successor_id: Optional[str] = None
    estate: EstateDisposition = field(default_factory=EstateDisposition)
    emergency: bool = False
    phase: str = "planned"  # planned | in_progress | complete

    def to_dict(self) -> Dict[str, Any]:
        return {
            "agent_id": self.agent_id,
            "reason": self.reason,
            "successor_id": self.successor_id,
            "estate": self.estate.to_dict(),
            "emergency": self.emergency,
            "phase": self.phase,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "DecommissionPlan":
        return cls(
            agent_id=d.get("agent_id", ""),
            reason=d.get("reason", "end_of_life"),
            successor_id=d.get("successor_id"),
            estate=EstateDisposition.from_dict(d.get("estate", {})),
            emergency=d.get("emergency", False),
            phase=d.get("phase", "planned"),
        )


def graceful_decommission(
    manager: LifecycleManager,
    plan: DecommissionPlan,
    initiator: Optional[Initiator] = None,
) -> DecommissionPlan:
    """Execute graceful (apoptotic) decommission (Section 9.1-9.2).

    Agent must be in Deprecated state. Follows full checklist:
    1. Verify obligations resolved
    2. Revoke credentials
    3. Dispose data
    4. Update registry and notifications
    """
    agent = manager.store.get_agent(plan.agent_id)
    if agent is None:
        raise LifecycleError(f"Agent not found: {plan.agent_id}")
    if agent.state != "deprecated":
        raise LifecycleError(
            f"Graceful decommission requires Deprecated state, "
            f"currently: {agent.state}. Deprecate the agent first."
        )

    plan.phase = "in_progress"

    event = LifecycleEvent(
        event_type="decommission",
        agent_id=plan.agent_id,
        agent_state_before="deprecated",
        agent_state_after="decommissioned",
        initiator=initiator or Initiator(),
        details={
            "reason": plan.reason,
            "successor_id": plan.successor_id,
            "estate_disposition": plan.estate.to_dict(),
            "decommission_type": "graceful_apoptosis",
        },
    )

    manager._run_pre_hooks(event, agent)

    agent.state = "decommissioned"
    agent.updated_at = _now_iso()
    agent.event_history.append(event.event_id)

    event.compute_hash()
    manager.store.append_event(event)
    manager.store.save_agent(agent)

    manager._run_post_hooks(event, agent)

    plan.phase = "complete"
    return plan


def emergency_decommission(
    manager: LifecycleManager,
    plan: DecommissionPlan,
    initiator: Optional[Initiator] = None,
) -> DecommissionPlan:
    """Execute emergency (necrotic) decommission (Section 9.4).

    Can be called from any non-terminal state. Credential revocation
    is immediate. Knowledge export may be skipped.
    """
    plan.emergency = True
    plan.phase = "in_progress"

    agent = manager.emergency_decommission(
        agent_id=plan.agent_id,
        reason=plan.reason,
        initiator=initiator,
    )

    plan.phase = "complete"
    return plan


def validate_decommission_checklist(
    plan: DecommissionPlan,
) -> List[str]:
    """Validate the decommission checklist (Section 9.2).

    Returns a list of items that are NOT yet completed.
    """
    issues: List[str] = []

    # Phase 1: Preparation
    if not plan.estate.counterparties_notified:
        issues.append("Counterparties not yet notified")

    # Phase 2: Credential Revocation
    cred = plan.estate.credentials
    if not cred.api_keys_revoked:
        issues.append("API keys not revoked")
    if not cred.oauth_tokens_invalidated:
        issues.append("OAuth tokens not invalidated")
    if not cred.service_accounts_deleted:
        issues.append("Service accounts not deleted")
    if not cred.trust_relationships_terminated:
        issues.append("Trust relationships not terminated")
    if not cred.certificates_revoked:
        issues.append("Certificates not revoked")
    if not cred.identity_key_archived:
        issues.append("Identity key not archived")

    # Phase 4: Registry and notification
    if not plan.estate.fleet_coordinator_notified:
        issues.append("Fleet coordinator not notified")

    return issues

File: typescript/src/decommission.ts
import {
  Initiator,
  LifecycleEvent,
  nowIso,
} from "./types";
import { LifecycleError, LifecycleManager } from "./lifecycle";

export interface CredentialRevocationData {
  all_api_keys_revoked: boolean;
  all_oauth_tokens_invalidated: boolean;
  all_service_accounts_deleted: boolean;
  all_trust_relationships_terminated: boolean;
  all_certificates_revoked: boolean;
  identity_key_archived: boolean;
}

export class CredentialRevocation {
  apiKeysRevoked: boolean;
  oauthTokensInvalidated: boolean;
  serviceAccountsDeleted: boolean;
  trustRelationshipsTerminated: boolean;
  certificatesRevoked: boolean;
  identityKeyArchived: boolean;

  constructor(opts?: Partial<{
    apiKeysRevoked: boolean;
    oauthTokensInvalidated: boolean;
    serviceAccountsDeleted: boolean;
    trustRelationshipsTerminated: boolean;
    certificatesRevoked: boolean;
    identityKeyArchived: boolean;
  }>) {
    this.apiKeysRevoked = opts?.apiKeysRevoked ?? false;
    this.oauthTokensInvalidated = opts?.oauthTokensInvalidated ?? false;
    this.serviceAccountsDeleted = opts?.serviceAccountsDeleted ?? false;
    this.trustRelationshipsTerminated = opts?.trustRelationshipsTerminated ?? false;
    this.certificatesRevoked = opts?.certificatesRevoked ?? false;
    this.identityKeyArchived = opts?.identityKeyArchived ?? false;
  }

  get allRevoked(): boolean {
    return (
      this.apiKeysRevoked &&
      this.oauthTokensInvalidated &&
      this.serviceAccountsDeleted &&
      this.trustRelationshipsTerminated &&
      this.certificatesRevoked &&
      this.identityKeyArchived
    );
  }

  toDict(): CredentialRevocationData {
    return {
      all_api_keys_revoked: this.apiKeysRevoked,
      all_oauth_tokens_invalidated: this.oauthTokensInvalidated,
      all_service_accounts_deleted: this.serviceAccountsDeleted,
      all_trust_relationships_terminated: this.trustRelationshipsTerminated,
      all_certificates_revoked: this.certificatesRevoked,
      identity_key_archived: this.identityKeyArchived,
    };
  }

  static fromDict(d: Partial<CredentialRevocationData>): CredentialRevocation {
    return new CredentialRevocation({
      apiKeysRevoked: d.all_api_keys_revoked,
      oauthTokensInvalidated: d.all_oauth_tokens_invalidated,
      serviceAccountsDeleted: d.all_service_accounts_deleted,
      trustRelationshipsTerminated: d.all_trust_relationships_terminated,
      certificatesRevoked: d.all_certificates_revoked,
      identityKeyArchived: d.identity_key_archived,
    });
  }
}

export interface DataDispositionData {
  operational_logs: string;
  memory_state: string;
  coc_chain: string;
  knowledge_artifacts: string;
}

export class DataDisposition {
  operationalLogs: string;
  memoryState: string;
  cocChain: string;
  knowledgeArtifacts: string;

  constructor(opts?: Partial<{
    operationalLogs: string;
    memoryState: string;
    cocChain: string;
    knowledgeArtifacts: string;
  }>) {
    this.operationalLogs = opts?.operationalLogs ?? "archived_90_days";
    this.memoryState = opts?.memoryState ?? "purged";
    this.cocChain = opts?.cocChain ?? "sealed_permanent";
    this.knowledgeArtifacts = opts?.knowledgeArtifacts ?? "transferred_to_fleet";
  }

  toDict(): DataDispositionData {
    return {
      operational_logs: this.operationalLogs,
      memory_state: this.memoryState,
      coc_chain: this.cocChain,
      knowledge_artifacts: this.knowledgeArtifacts,
    };
  }

  static fromDict(d: Partial<DataDispositionData>): DataDisposition {
    return new DataDisposition({
      operationalLogs: d.operational_logs,
      memoryState: d.memory_state,
      cocChain: d.coc_chain,
      knowledgeArtifacts: d.knowledge_artifacts,
    });
  }
}

export interface EstateDispositionData {
  obligations: { transferred: number; terminated: number };
  credentials: CredentialRevocationData;
  data: DataDispositionData;
  notifications: {
    counterparties_notified: boolean;
    fleet_coordinator_notified: boolean;
  };
}

export class EstateDisposition {
  obligationsTransferred: number;
  obligationsTerminated: number;
  credentials: CredentialRevocation;
  data: DataDisposition;
  counterpartiesNotified: boolean;
  fleetCoordinatorNotified: boolean;

  constructor(opts?: Partial<{
    obligationsTransferred: number;
    obligationsTerminated: number;
    credentials: CredentialRevocation;
    data: DataDisposition;
    counterpartiesNotified: boolean;
    fleetCoordinatorNotified: boolean;
  }>) {
    this.obligationsTransferred = opts?.obligationsTransferred ?? 0;
    this.obligationsTerminated = opts?.obligationsTerminated ?? 0;
    this.credentials = opts?.credentials ?? new CredentialRevocation();
    this.data = opts?.data ?? new DataDisposition();
    this.counterpartiesNotified = opts?.counterpartiesNotified ?? false;
    this.fleetCoordinatorNotified = opts?.fleetCoordinatorNotified ?? false;
  }

  toDict(): EstateDispositionData {
    return {
      obligations: {
        transferred: this.obligationsTransferred,
        terminated: this.obligationsTerminated,
      },
      credentials: this.credentials.toDict(),
      data: this.data.toDict(),
      notifications: {
        counterparties_notified: this.counterpartiesNotified,
        fleet_coordinator_notified: this.fleetCoordinatorNotified,
      },
    };
  }

  static fromDict(d: Partial<EstateDispositionData>): EstateDisposition {
    const obligations = d.obligations ?? {};
    const notif = d.notifications ?? {};
    return new EstateDisposition({
      obligationsTransferred: (obligations as Record<string, number>).transferred,
      obligationsTerminated: (obligations as Record<string, number>).terminated,
      credentials: CredentialRevocation.fromDict(d.credentials ?? {}),
      data: DataDisposition.fromDict(d.data ?? {}),
      counterpartiesNotified: (notif as Record<string, boolean>).counterparties_notified,
      fleetCoordinatorNotified: (notif as Record<string, boolean>).fleet_coordinator_notified,
    });
  }
}

export interface DecommissionPlanData {
  agent_id: string;
  reason: string;
  successor_id: string | null;
  estate: EstateDispositionData;
  emergency: boolean;
  phase: string;
}

export class DecommissionPlan {
  agentId: string;
  reason: string;
  successorId: string | null;
  estate: EstateDisposition;
  emergency: boolean;
  phase: string;

  constructor(opts?: Partial<{
    agentId: string;
    reason: string;
    successorId: string | null;
    estate: EstateDisposition;
    emergency: boolean;
    phase: string;
  }>) {
    this.agentId = opts?.agentId ?? "";
    this.reason = opts?.reason ?? "end_of_life";
    this.successorId = opts?.successorId ?? null;
    this.estate = opts?.estate ?? new EstateDisposition();
    this.emergency = opts?.emergency ?? false;
    this.phase = opts?.phase ?? "planned";
  }

  toDict(): DecommissionPlanData {
    return {
      agent_id: this.agentId,
      reason: this.reason,
      successor_id: this.successorId,
      estate: this.estate.toDict(),
      emergency: this.emergency,
      phase: this.phase,
    };
  }

  static fromDict(d: Partial<DecommissionPlanData>): DecommissionPlan {
    return new DecommissionPlan({
      agentId: d.agent_id,
      reason: d.reason,
      successorId: d.successor_id,
      estate: EstateDisposition.fromDict(d.estate ?? {}),
      emergency: d.emergency,
      phase: d.phase,
    });
  }
}

export function gracefulDecommission(
  manager: LifecycleManager,
  plan: DecommissionPlan,
  initiator?: Initiator,
): DecommissionPlan {
  const agent = manager.store.getAgent(plan.agentId);
  if (agent === null) {
    throw new LifecycleError(`Agent not found: ${plan.agentId}`);
  }
  if (agent.state !== "deprecated") {
    throw new LifecycleError(
      `Graceful decommission requires Deprecated state, ` +
        `currently: ${agent.state}. Deprecate the agent first.`,
    );
  }

  plan.phase = "in_progress";

  const event = new LifecycleEvent({
    eventType: "decommission",
    agentId: plan.agentId,
    agentStateBefore: "deprecated",
    agentStateAfter: "decommissioned",
    initiator: initiator ?? new Initiator(),
    details: {
      reason: plan.reason,
      successor_id: plan.successorId,
      estate_disposition: plan.estate.toDict(),
      decommission_type: "graceful_apoptosis",
    },
  });

  manager._runPreHooks(event, agent);

  agent.state = "decommissioned";
  agent.updatedAt = nowIso();
  agent.eventHistory.push(event.eventId);

  event.computeHash();
  manager.store.appendEvent(event);
  manager.store.saveAgent(agent);

  manager._runPostHooks(event, agent);

  plan.phase = "complete";
  return plan;
}

export function emergencyDecommission(
  manager: LifecycleManager,
  plan: DecommissionPlan,
  initiator?: Initiator,
): DecommissionPlan {
  plan.emergency = true;
  plan.phase = "in_progress";

  manager.emergencyDecommission(
    plan.agentId,
    plan.reason,
    initiator,
  );

  plan.phase = "complete";
  return plan;
}

export function validateDecommissionChecklist(
  plan: DecommissionPlan,
): string[] {
  const issues: string[] = [];

  if (!plan.estate.counterpartiesNotified) {
    issues.push("Counterparties not yet notified");
  }

  const cred = plan.estate.credentials;
  if (!cred.apiKeysRevoked) issues.push("API keys not revoked");
  if (!cred.oauthTokensInvalidated)
    issues.push("OAuth tokens not invalidated");
  if (!cred.serviceAccountsDeleted)
    issues.push("Service accounts not deleted");
  if (!cred.trustRelationshipsTerminated)
    issues.push("Trust relationships not terminated");
  if (!cred.certificatesRevoked) issues.push("Certificates not revoked");
  if (!cred.identityKeyArchived) issues.push("Identity key not archived");

  if (!plan.estate.fleetCoordinatorNotified) {
    issues.push("Fleet coordinator not notified");
  }

  return issues;
}
